README.md
# Barely JSON

[![GitHub Actions badge](https://github.com/torfsen/barely_json/actions/workflows/test.yml/badge.svg)](https://github.com/torfsen/barely_json/actions/workflows/test.yml)

*A Python package for parsing data that only looks like JSON*

    from barely_json import parse
    print(parse('[what is this?, , {perhaps, json: no}]'))

    # Prints ['what is this?', '', {'perhaps': '', 'json': False}]

Quite a bit of data looks like JSON at a first glance but turns out not to comply completely with the JSON specification -- often because the exporting software is broken, but sometimes simply because the format was never intended to be JSON in the first place.

No matter how you ended up with the data, now you want to parse it! However, most JSON parsers are pretty strict, so you're out of luck with your JSON-esque mess.

That's where *Barely JSON* steps in and tries to parse anything that remotely looks like JSON. In addition to the pure parsing, *Barely JSON* will also try to post-process your data into suitable Python types even if your data provider uses, for example, `on` and `off` as boolean literals.


# Installation

The supported Python versions are 3.7 and later.

    pip install barely_json


# Usage

The main routine is `parse`:

    > from barely_json import parse
    > parse("[NaN, , {state: off, where's my value?}, NULL]")

    [nan, '', {'state': False, "where's my value?": ''}, None]

As you can see, `parse` by default tries to convert values that are illegal in JSON into hopefully appropriate Python types, which often works well. But sometimes that's not what you want, so you can disable the auto-conversion:

    > parse("[NaN, , {state: off, where's my value?}, NULL]", resolver=None)

    [<IllegalValue 'NaN'>,
     <IllegalValue ''>,
     {<IllegalValue 'state'>: <IllegalValue 'off'>,
      <IllegalValue "where's my value?">: <IllegalValue ''>},
     <IllegalValue 'NULL'>]

In that case any value that's illegal or missing is wrapped in an instance of a special `IllegalValue` class. You can also provide your own resolver for illegal values, which is simply a callback that maps strings to arbitrary values:

    > from barely_json import default_resolver
    >
    > def my_resolver(text):
    >     if text.lower() == 'one':
    >         return 1
    >     return default_resolver(text)
    >
    > parse('[one, FALSE]', resolver=my_resolver)

    [1, False]

When writing your own resolver it's often handy to fall back to`default_resolver` after you've handled your special cases.


# Change Log

See `CHANGELOG.md`.


# License

Distributed under the MIT license. See the file `LICENSE` for details.


# Contributors

* [@torfsen](https://github.com/torfsen)
* [@tusharmakkar08](https://github.com/tusharmakkar08)
* [@gvx](https://github.com/gvx)


# Development

Clone the repository:

    git clone https://github.com/torfsen/barely_json.git
    cd barely_json

Install the development dependencies

    pip install -r requirements-dev.txt

Run the tests:

    tox

For pull requests, the tests are run using GitHub actions.


[virtualenv]: https://virtualenv.pypa.io
[tox]: https://tox.readthedocs.io


CHANGELOG.md
# Change Log for Barely JSON

The format of this file is based on [Keep a Changelog] and this project adheres to [Semantic Versioning].


## [1.0.0] (2023-07-29)

### Fixed

- `IllegalValue.__str__` was broken (reported and fixed by [@tusharmakkar08] in [#4])

### Added

- Support for Python 3.7, 3.8, 3.9, 3.10, and 3.11

### Removed

- Support for Python 2

### Changed

- Code cleanup (contributed by [@gvx] in [#3])


## [0.1.1] (2017-05-12)

### Fixed

- Floats were not parsed correctly ([#1])
- String escapes were not parsed correctly ([#2])


## [0.1.0] (2017-05-12)

### Added

- Support for parsing strings
- Support for custom resolvers for illegal values


[Keep a Changelog]: http://keepachangelog.com/
[Semantic Versioning]: http://semver.org/

[Unreleased]: https://github.com/torfsen/barely_json/compare/v1.0.0...master
[1.0.0]: https://github.com/torfsen/barely_json/compare/v0.1.1...v1.0.0
[0.1.1]: https://github.com/torfsen/barely_json/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/torfsen/barely_json/commits/v0.1.0

[@gvx]: https://github.com/gvx
[@tusharmakkar08]: https://github.com/tusharmakkar08

[#1]: https://github.com/torfsen/barely_json/issues/1
[#2]: https://github.com/torfsen/barely_json/issues/2
[#3]: https://github.com/torfsen/barely_json/pull/3
[#4]: https://github.com/torfsen/barely_json/pull/4


tests/test_barely_json.py
#!/usr/bin/env python

import math

from barely_json import *

empty = IllegalValue('')

STRING_ESCAPE_TESTS = [
    r'\t', '\t',
    r'\\', '\\',
    r'\\\\', '\\\\',
    r'\\\\\\', '\\\\\\',
    r'\"', '"',
    r'\\\"', '\\"',
    r'\\\\\"', '\\\\"',
    r'\'', "'",
    r'\a', '\a',
    r'\b', '\b',
    r'\f', '\f',
    r'\n', '\n',
    r'\r', '\r',
    r'\t', '\t',
    r'\v', '\v',
    r'\o1', r'\o1',
    r'\o12', r'\o12',
    r'\o123', r'\o123',
    r'\xef', '\xef',
    r'\xEF', '\xef',
    r'\N{SNOWMAN}', '\N{SNOWMAN}',
    r'\u2603', '\u2603',
    r'\U00002603', '\U00002603',
]


def pairwise(iterable):
    '''
    s -> (s0, s1), (s2, s3), (s4, s5), ...
    '''
    a = iter(iterable)
    return zip(a, a)


class TestParse:
    @staticmethod
    def check(*args, **kwargs):
        resolver = kwargs.pop('resolver', None)
        for code, expected in pairwise(args):
            assert parse(code, resolver=resolver) == expected

    def test_empty_list(self):
        self.check(
            '[]', [],
            '[ ]', [],
            '[\n]', [],
            '[\t]', [],
        )

    def test_empty_list_element(self):
        self.check(
            '[,]', [empty, empty],
            '[,,]', [empty, empty, empty],
            '[1,]', [1, empty],
            '[1,,]', [1, empty, empty],
            '[,1]', [empty, 1],
            '[,,1]', [empty, empty, 1],
            '[1,,2]', [1, empty, 2],
            '[1,,2,]', [1, empty, 2, empty],
            '[1,,,2]', [1, empty, empty, 2],
        )

    def test_empty_dict(self):
        self.check(
            '{}', {},
            '{ }', {},
            '{\n}', {},
            '{\t}', {},
        )

    def test_dict(self):
        self.check(
            '{1: 2}', {IllegalValue('1'): 2},
            '{1: 2, 3: 4}', {IllegalValue('1'): 2, IllegalValue('3'): 4},
            "{'1': 2}", {IllegalValue("'1'"): 2},
            '{"1": 2}', {'1': 2},
            '''{a b ":" ':' 2 3 : 1}''', {IllegalValue('''a b ":" ':' 2 3'''): 1},
        )

    def test_empty_dict_element(self):
        self.check(
            '{,}', {},
            '{,,}', {},
            '{,,,}', {},
            '{"a":1,}', {'a': 1},
            '{"a":1,,}', {'a': 1},
            '{,,"a":1}', {'a': 1},
            '{,"a":1,}', {'a': 1},
            '{"a":1,,"b":2}', {'a': 1, 'b': 2},
            '{"a":1,,,"b":2}', {'a': 1, 'b': 2},
        )

    def test_missing_dict_value(self):
        self.check(
            '{"a"}', {'a': empty},
            '{"a", "b": 1}', {'a': empty, 'b': 1},
            '{"a": 1, "b"}', {'a': 1, 'b': empty},
            '{"a":}', {'a': empty},
            '{"a":, "b": 1}', {'a': empty, 'b': 1},
            '{"a": 1, "b":}', {'a': 1, 'b': empty},
        )

    def test_numbers(self):
        self.check(
            '0', 0,
            '-0', 0,
            '+0', 0,
            '1', 1,
            '+1', 1,
            '-1', -1,
            '10', 10,
            '+10', 10,
            '-10', -10,
            '0.123', 0.123,
            '+0.123', 0.123,
            '-0.123', -0.123,
            '.123', .123,
            '+.123', .123,
            '-.123', -.123,
            '1.', 1.0,
            '+1.', 1.0,
            '-1.', -1.0,
            '123.e10', 123e10,
            '-123.E-3', -123e-3,
        )

    def test_string_escapes(self):
        cases = (
            ('"{}"'.format(x), y) for x, y in pairwise(STRING_ESCAPE_TESTS)
        )
        self.check(*(x for case in cases for x in case))


class TestDefaultResolver:
    @staticmethod
    def check(*args):
        for value, expected in pairwise(args):
            assert default_resolver(value) == expected

    def test_empty(self):
        self.check('', '')

    def test_boolean(self):
        self.check(
            'TRUE', True,
            'true', True,
            'YES', True,
            'yes', True,
            'ON', True,
            'on', True,
            'FALSE', False,
            'false', False,
            'NO', False,
            'no', False,
            'OFF', False,
            'off', False,
        )

    def test_null(self):
        self.check(
            'NULL', None,
            'null', None,
            'NONE', None,
            'none', None,
        )

    def test_float(self):
        self.check(
            'INF', float('inf'),
            'inf', float('inf'),
            '-INF', float('-inf'),
            '-inf', float('-inf'),
            '0', 0,
            '1', 1,
            '-1', -1,
            '0.04', 0.04,
            '-21.432', -21.432,
            '2.4e-3', 2.4e-3,
            '-34.12E9', -34.12e9,
        )
        for value in ['nan', 'NaN', 'NAN']:
            assert math.isnan(default_resolver(value))

    def test_unknown(self):
        self.check(
            'foo', 'foo',
            'FOO', 'FOO',
        )

    def test_string_escapes(self):
        self.check(*STRING_ESCAPE_TESTS)


class TestResolve:
    @staticmethod
    def check(*args, **kwargs):
        resolver = kwargs.pop('resolver', default_resolver)
        for value, expected in pairwise(args):
            assert resolve(value, resolver=resolver) == expected

    def test_list(self):
        self.check(
            [], [],
            [1, 'foo', True], [1, 'foo', True],
            [empty, IllegalValue('NO')], ['', False],
        )

    def test_dict(self):
        self.check(
            {}, {},
            {'a': 1, 'b': True, 'c': None}, {'a': 1, 'b': True, 'c': None},
            {'a': empty}, {'a': ''},
            {'a': IllegalValue('false')}, {'a': False},
            {IllegalValue('YES'): IllegalValue('no')}, {True: False},
        )

    def test_scalar(self):
        self.check(
            1, 1,
            'foo', 'foo',
            empty, '',
            IllegalValue('YES'), True,
        )

    def test_nested(self):
        self.check(
            [empty, {IllegalValue('yes'): [IllegalValue('no')]}],
            ['', {True: [False]}]
        )

    def test_custom_resolver(self):
        def resolver(value):
            if value == 'one':
                return 1
            return value

        self.check(
            IllegalValue('one'), 1,
            IllegalValue('true'), 'true',
            [IllegalValue('one')], [1],
            [IllegalValue('true')], ['true'],
            {IllegalValue('one'): IllegalValue('one')}, {1: 1},
            {IllegalValue('true'): IllegalValue('true')}, {'true': 'true'},
            resolver=resolver
        )


class TestIllegalValue:
    def test_str_illegal(self):
        assert str(IllegalValue('one')) == 'one'


barely_json/grammar.py
#!/usr/bin/env python
# encoding: utf-8
# __author__ = 'tusharmakkar08'

import codecs
import re

from pyparsing import *

from barely_json.illegal_value import IllegalValue


def to_float(t):
    '''
    Convert a string that looks like a float to a float.
    '''
    return float(re.sub(r'\s*', '', t[0]))


def to_int(t):
    '''
    Convert a string that looks like an int to an int.
    '''
    return int(re.sub(r'\s*', '', t[0]))


# From http://stackoverflow.com/a/24519338/857390

ESCAPE_SEQUENCE_RE = re.compile(r'''
    ( \\U........      # 8-digit hex escapes
    | \\u....          # 4-digit hex escapes
    | \\x..            # 2-digit hex escapes
    | \\[0-7]{1,3}     # Octal escapes
    | \\N\{[^}]+\}     # Unicode characters by name
    | \\[\\'"abfnrtv]  # Single-character escapes
    )''', re.UNICODE | re.VERBOSE)


def decode_escapes(s):
    '''
    Decode string escape sequences.
    '''
    def decode_match(match):
        return codecs.decode(match.group(0), 'unicode-escape')
    result = ESCAPE_SEQUENCE_RE.sub(decode_match, s)
    return result


L_BRACKET, R_BRACKET, L_BRACE, R_BRACE, COLON, COMMA = map(Suppress, '[]{}:,')

integrated_grammar_value = Forward()

null = Keyword('null').setParseAction(lambda: None)

true = Keyword('true').setParseAction(lambda: True)
false = Keyword('false').setParseAction(lambda: False)

int_ = Regex(r'[+-]?\s*\d+').setParseAction(to_int)

FLOAT_RE = r'[+-]?\s*(0(\.\d*)?|([1-9]\d*\.?\d*)|(\.\d+))([Ee][+-]?\d+)?'

float_ = Regex(FLOAT_RE).setParseAction(to_float)

# PyParsing's QuotedString preprocesses some escape sequences incorrectly,
# so we're doing things manually.
string_ = Regex(r'".*?(?<!\\)(\\\\)*"')\
            .setParseAction(lambda t: decode_escapes(t[0][1:-1]))

empty = Empty().setParseAction(lambda: IllegalValue(''))

# Special case the empty list to avoid [IllegalValue('')]
list_ = (
    (L_BRACKET + R_BRACKET).setParseAction(lambda: [[]]) |
    Group(L_BRACKET + Optional(delimitedList(integrated_grammar_value ^ empty)) + R_BRACKET)
)

illegal = Combine(OneOrMore(quotedString ^ Regex(r'[^,{}[\]]')))\
                .setParseAction(lambda t: IllegalValue(t[0].strip()))
# Like ``illegal`` but doesn't allow colons
illegal_key = Combine(OneOrMore(quotedString ^ Regex(r'[^:,{}[\]]')))\
                .setParseAction(lambda t: IllegalValue(t[0].strip()))

dict_key = string_ | illegal_key
dict_value = Optional(integrated_grammar_value, default=IllegalValue(''))
dict_item = Group(dict_key + Optional(COLON + dict_value,
                  default=IllegalValue('')))
dict_content = delimitedList(dict_item ^ Empty())\
                .setParseAction(lambda t: dict(t.asList()))
dict_ = L_BRACE + dict_content + R_BRACE

integrated_grammar_value << (dict_ | list_ | float_ | int_ | string_ |
                             null | true | false | illegal)


barely_json/illegal_value.py
# __author__ = 'tusharmakkar08'


class IllegalValue(object):
    '''
    A value that is illegal in JSON.

    ``parse`` wraps anything that isn't standard JSON into an
    ``IllegalValue`` instance. By default, these are then automatically
    resolved into standard Python types via ``resolve``. However, if you
    pass ``resolver=None`` to ``parse`` and your input contains illegal
    values then your output will contain instances of this class.

    The part of the source that is represented by this instance is
    stored in the ``source`` attribute. That may be the empty string in
    cases like ``[1, , 2]``.
    '''

    def __init__(self, source):
        '''
        Constructor.

        ``source`` is a string.
        '''
        self.source = source

    def __str__(self):
        return self.source

    def __repr__(self):
        return '<{} {!r}>'.format(self.__class__.__name__, self.source)

    def __eq__(self, other):
        return (isinstance(other, self.__class__) and
                other.source == self.source)

    def __hash__(self):
        return hash(self.source)


barely_json/__init__.py
#!/usr/bin/env python

'''
A very forgiving JSON parser.
'''

import codecs
import re

from pyparsing import *


__version__ = '1.0.0'


class IllegalValue:
    '''
    A value that is illegal in JSON.

    ``parse`` wraps anything that isn't standard JSON into an
    ``IllegalValue`` instance. By default, these are then automatically
    resolved into standard Python types via ``resolve``. However, if you
    pass ``resolver=None`` to ``parse`` and your input contains illegal
    values then your output will contain instances of this class.

    The part of the source that is represented by this instance is
    stored in the ``source`` attribute. That may be the empty string in
    cases like ``[1, , 2]``.
    '''
    def __init__(self, source):
        '''
        Constructor.

        ``source`` is a string.
        '''
        self.source = source

    def __str__(self):
        return self.source

    def __repr__(self):
        return '<{} {!r}>'.format(self.__class__.__name__, self.source)

    def __eq__(self, other):
        return (isinstance(other, self.__class__) and
                other.source == self.source)

    def __hash__(self):
        return hash(self.source)


def to_float(t):
    '''
    Convert a string that looks like a float to a float.
    '''
    return float(re.sub(r'\s*', '', t[0]))


def to_int(t):
    '''
    Convert a string that looks like an int to an int.
    '''
    return int(re.sub(r'\s*', '', t[0]))


# From http://stackoverflow.com/a/24519338/857390

ESCAPE_SEQUENCE_RE = re.compile(r'''
    ( \\U........      # 8-digit hex escapes
    | \\u....          # 4-digit hex escapes
    | \\x..            # 2-digit hex escapes
    | \\[0-7]{1,3}     # Octal escapes
    | \\N\{[^}]+\}     # Unicode characters by name
    | \\[\\'"abfnrtv]  # Single-character escapes
    )''', re.UNICODE | re.VERBOSE)


def decode_escapes(s):
    '''
    Decode string escape sequences.
    '''
    def decode_match(match):
        return codecs.decode(match.group(0), 'unicode-escape')
    result = ESCAPE_SEQUENCE_RE.sub(decode_match, s)
    return result


L_BRACKET, R_BRACKET, L_BRACE, R_BRACE, COLON, COMMA = map(Suppress, '[]{}:,')

value = Forward()

null = Keyword('null').setParseAction(lambda: None)

true = Keyword('true').setParseAction(lambda: True)
false = Keyword('false').setParseAction(lambda: False)

int_ = Regex(r'[+-]?\s*\d+').setParseAction(to_int)

FLOAT_RE = r'[+-]?\s*(0(\.\d*)?|([1-9]\d*\.?\d*)|(\.\d+))([Ee][+-]?\d+)?'

float_ = Regex(FLOAT_RE).setParseAction(to_float)

# PyParsing's QuotedString preprocesses some escape sequences incorrectly,
# so we're doing things manually.
string_ = Regex(r'".*?(?<!\\)(\\\\)*"')\
            .setParseAction(lambda t: decode_escapes(t[0][1:-1]))

empty = Empty().setParseAction(lambda: IllegalValue(''))

# Special case the empty list to avoid [IllegalValue('')]
list_ = (
    (L_BRACKET + R_BRACKET).setParseAction(lambda: [[]]) |
    Group(L_BRACKET + Optional(delimitedList(value ^ empty)) + R_BRACKET)
)

illegal = Combine(OneOrMore(quotedString ^ Regex(r'[^,{}[\]]')))\
                .setParseAction(lambda t: IllegalValue(t[0].strip()))
# Like ``illegal`` but doesn't allow colons
illegal_key = Combine(OneOrMore(quotedString ^ Regex(r'[^:,{}[\]]')))\
                .setParseAction(lambda t: IllegalValue(t[0].strip()))

dict_key = string_ | illegal_key
dict_value = Optional(value, default=IllegalValue(''))
dict_item = Group(dict_key + Optional(COLON + dict_value,
                  default=IllegalValue('')))
dict_content = delimitedList(dict_item ^ Empty())\
                .setParseAction(lambda t: dict(t.asList()))
dict_ = L_BRACE + dict_content + R_BRACE

value << (dict_ | list_ | float_ | int_ | string_ |
          null | true | false | illegal)


def default_resolver(value):
    '''
    Default resolver for illegal values.

    Case-insensitively resolves

    - ``'true'``, ``'yes'``, and ``'on'`` to ``True``

    - ``'false'``, ``'no'``, and ``'off'`` to ``False``

    - ``'null'`` and ``'none'`` to ``None``

    - ``'inf'``, ``'-inf'``, and ``'nan'`` to their ``float``
      counterparts

    All other strings are returned unchanged.
    '''
    low = value.lower()
    if low in ['true', 'yes', 'on']:
        return True
    if low in ['false', 'no', 'off']:
        return False
    if low in ['null', 'none']:
        return None
    try:
        # Handles +/- NaN/Inf
        return float(re.sub(r'\s+', '', low))
    except ValueError:
        pass
    return decode_escapes(value)


def resolve(data, resolver=default_resolver):
    '''
    Recursively resolve illegal values.

    ``data`` is a potentially nested Python value as returned by
    ``parse``.

    ``resolver`` is a function that maps strings to arbitrary values.

    All instances of ``IllegalValue`` in ``data`` are replaced by the
    result of feeding them into ``resolver``.
    '''
    if isinstance(data, list):
        return [resolve(item, resolver) for item in data]
    if isinstance(data, dict):
        return {resolve(key, resolver): resolve(value, resolver) for key, value in data.items()}
    if isinstance(data, IllegalValue):
        return resolver(data.source)
    return data


def parse(s, resolver=default_resolver):
    '''
    Parse a string that contains barely JSON.

    When values that are illegal in JSON are encountered then they are
    by default heuristicaly resolved to a suitable Python type (see
    ``resolve`` and ``default_resolver``). Set ``resolver`` to a custom
    callback or to a falsy value to modify or disable that mechanism.
    '''
    parsed = value.parseString(s, parseAll=True)
    data = parsed.asList()[0]
    if resolver:
        data = resolve(data, resolver=resolver)
    return data
